doc/stdenv/platform-notes.chapter.md

```diff
@@ -46,22 +46,195 @@ Some common issues when packaging software for Darwin:
+
+  - If a package specifies a range of supported SDK versions _and_ a minimum supported version, assume the package is using availability checks to support the indicated minimum version. Add the highest supported SDK and a `darwinMinVersionHook` set to the minimum version supported by the upstream package.
+
+  Warning: Avoid using newer SDKs than an upstream package supports. When a binary is linked on Darwin, the SDK version used to build it is recorded in the binary. Runtime behavior can vary based on the SDK version, which may work fine but can also result in unexpected behavior or crashes when building with an unsupported SDK.
+
+  ```nix
+  stdenv.mkDerivation {
+    name = "foo-1.2.3";
+    # ...
+    buildInputs = [ apple-sdk_15 (darwinMinVersionHook "10.15") ]; # Upstream builds with the 15.0 SDK but supports 10.15.
+  }
+  ```
+
+- Libraries that require a minimum version can propagate an appropriate SDK and `darwinMinVersionHook`. Derivations using that library will automatically use an appropriate SDK and minimum version. Even if the library builds with a newer SDK, it should propagate the minimum supported SDK. Derivations that need a newer SDK can add it to their `buildInputs`.
+
+  ```nix
+  stdenv.mkDerivation {
+    name = "libfoo-1.2.3";
+    # ...
+    buildInputs = [ apple-sdk_15 ]; # Upstream builds with the 15.0 SDK but supports 10.15.
+    propagatedBuildInputs = [ apple-sdk_10_15 (darwinMinVersionHook "10.15") ];
+  }
+  # ...
+  stdenv.mkDerivation {
+    name = "bar-1.2.3";
+    # ...
+    buildInputs = [ libfoo ]; # Builds with the 10.15 SDK
+  }
+  # ...
+  stdenv.mkDerivation {
+    name = "baz-1.2.3";
+    # ...
+    buildInputs = [ apple-sdk_12 libfoo ]; # Builds with the 12.3 SDK
+  }
+  ```
+
+- Many of the SDK libraries are built from source (typically corresponding to the source releases for the latest release of macOS). Several of these are propagated to your package automatically. They can be accessed via the `darwin` package set.
+
+  - libiconv
+  - libresolv
+  - libsbuf
+
+  Other common libraries are available in Darwin-specific versions with modifications from Apple. Note that these packages may be made the default on Darwin in the future.
+
+  - ICU (compatible with the top-level icu package, but it also provides `libicucore.B.dylib` with an ABI compatible with the Darwin system version)
+  - libpcap (compatible with the top-level libpcap, but it includes Darwin-specific extensions)
+
+- The legacy SDKs packages are still available in the `darwin` package set under their existing names, but all packages in these SDKs (frameworks, libraries, etc) have been replaced with stubs containing a readme advising you to read this documentation for how to update your derivation to use the new pattern. Note that the new SDK pattern uses the name `apple-sdk` to better align with nixpkgs naming conventions.
```

Comment:
Probably more helpful to focus on the migration guide part of this than the details of how we’ve implemented the shims.

Reply:
I wanted to make it clear these legacy frameworks do nothing, and that’s expected behavior.